staging/src/k8s.io/apiserver/pkg/storage/feature/feature_support_checker.go

```diff
@@ -43,84 +46,95 @@ type FeatureSupportChecker interface {
 	switch feature {
 	case storage.RequestWatchProgress:
 		f.lock.Lock()
 		defer f.lock.Unlock()
 
-		return ptr.Deref(f.progressNotifySupported, false), nil
+		return ptr.Deref(f.progressNotifySupported, false)
 	default:
-		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+		runtime.HandleError(fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature))
+		return false
 	}
 }
 
 // CheckClient accepts client and calculate the support per endpoint and caches it.
-// It will return at any point if error happens or one endpoint is not supported.
-func (f *defaultFeatureSupportChecker) CheckClient(ctx context.Context, c client, feature storage.Feature) error {
+func (f *defaultFeatureSupportChecker) CheckClient(ctx context.Context, c client, feature storage.Feature) {
 	switch feature {
 	case storage.RequestWatchProgress:
-		return f.clientSupportsRequestWatchProgress(ctx, c)
+		f.checkClient(ctx, c)
 	default:
-		return fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
-
+		runtime.HandleError(fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature))
 	}
 }
 
-func (f *defaultFeatureSupportChecker) clientSupportsRequestWatchProgress(ctx context.Context, c client) error {
+func (f *defaultFeatureSupportChecker) checkClient(ctx context.Context, c client) {
+	// start with 10 ms, multiply by 2 each step, until 15 s and stays on 15 seconds.
+	delayFunc := wait.Backoff{
+		Duration: 10 * time.Millisecond,
+		Cap:      15 * time.Second,
+		Factor:   2.0,
+		Steps:    11}.DelayFunc()
 	f.lock.Lock()
 	defer f.lock.Unlock()
-
 	for _, ep := range c.Endpoints() {
-		supported, err := f.supportsProgressNotifyEndpointLocked(ctx, c, ep)
-		if err != nil {
-			return err
-		}
-		if !supported {
-			f.progressNotifySupported = ptr.To(false)
-			return nil
+		if _, found := f.checkingEndpoint[ep]; found {
+			continue
 		}
+		f.checkingEndpoint[ep] = struct{}{}
+		go func(ep string) {
+			defer runtime.HandleCrash()
+			err := delayFunc.Until(ctx, true, true, func(ctx context.Context) (done bool, err error) {
+				internalErr := f.clientSupportsRequestWatchProgress(ctx, c, ep)
+				return internalErr == nil, nil
+			})
+			if err != nil {
+				klog.ErrorS(err, "Failed to check if RequestWatchProgress is supported by etcd after retrying")
+			}
+		}(ep)
 	}
-	if f.progressNotifySupported == nil && len(c.Endpoints()) > 0 {
-		f.progressNotifySupported = ptr.To(true)
-	}
-	return nil
 }
 
-func (f *defaultFeatureSupportChecker) supportsProgressNotifyEndpointLocked(ctx context.Context, c client, ep string) (bool, error) {
-	if supported, ok := f.progresNotifyEndpointCache[ep]; ok {
-		return supported, nil
-	}
-
+func (f *defaultFeatureSupportChecker) clientSupportsRequestWatchProgress(ctx context.Context, c client, ep string) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
```

Comment:
You don't want to hold a lock when sending an external call (to etcd).

The lock should be taken somewhere below.

Reply:
Thanks, done.